File: 0075-sort-colors/0075-sort-colors.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
// var sortColors = function(nums) {
//     //O(nlogn)으로 푸는 단순한 방법은 정렬
//     return nums.sort();
// };

//TODO O(n)으로 푸는방법
var sortColors = function(nums){
    let l = 0; //0 만 담는 인덱스
    let r = nums.length - 1; //2만 담는 인덱스

    let curIndex = 0;
    while(curIndex <= r){
        if(nums[curIndex] == 0){
            [nums[curIndex],nums[l]] = [nums[l],nums[curIndex]];
            l++
            curIndex++
        } else if(nums[curIndex] == 2){
            [nums[r], nums[curIndex]] = [nums[curIndex], nums[r]];
            r--
        } else {
            curIndex++;
        }
    }

}

File: 0075-sort-colors/0075-sort-colors.go
func sortColors(nums []int)  {
    l, r := 0, len(nums) - 1;
    curIndex :=0

    //l 이 0을 담을 인덱스 r이 2를 담을 인덱스

    for curIndex <= r {
        if nums[curIndex] == 0 {
            //0이라면 l이랑 스왑
            nums[l], nums[curIndex] = nums[curIndex], nums[l];
            l++;
            curIndex++;
        } else if nums[curIndex] == 2 {
            nums[r], nums[curIndex] = nums[curIndex], nums[r];
            r-- //스왑 한 뒤 1이나 0이 될 수도 있으므로 
        } else {
            curIndex++; //1은 그냥 넘어가기
        }
    }
}